### PR TITLE
Make the `semgrep ci` tests for github envs more accurate

### PR DESCRIPTION
**Pull request by jbergler**
The mocked env didn't accurately reflect the behavior of github actions which meant that the snapshots weren't accurately reflecting what we send to semgrep.dev in a scan. Before I make changes to the behavior I thought it appropriate to update these tests so that the diffs in future PRs are easier to review/understand.

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: cli/src/semgrep/commands/ci.py [hunk 1]
```diff
@@ -110,7 +110,7 @@ def fix_head_if_github_action(metadata: GitMeta) -> None:
     logger.info(f"Not on head ref: {metadata.head_branch_hash}; checking that out now.")
     git_check_output(["git", "checkout", metadata.head_branch_hash])
 
-    atexit.register(git_check_output, ["git", "checkout", stashed_rev])
+    atexit.register(git_check_output, ["git", "checkout", stashed_rev], os.getcwd())
```
Comment (jbergler) [inline cli/src/semgrep/commands/ci.py:113]: I think this is an edge case that only shows up in testing because of how we invoke semgrep during the tests but by the time the atexit runs the working directory has changed, we can pass the current value to git_check_output to ensure this behaves consistently and reduces some noise in the test env.

File: cli/tests/e2e/test_ci.py [hunk 2]
```diff
@@ -195,7 +223,7 @@ def git_tmp_path_with_commit(monkeypatch, tmp_path, mocker):
         check=True,
         capture_output=True,
     )
-    subprocess.run(["git", "pull", "origin"])
+    subprocess.run(["git", "fetch", "origin"])
```
Comment (jbergler) [inline cli/tests/e2e/test_ci.py:226]: `pull` throws warnings here because there is no branch checked out in this new repo we created, `fetch` is less distracting for people trying to debug because it removes the meaningless error.

**Comment by aryx** [comment]
Not super familiar with this code, but LGTM!

File: cli/tests/e2e/test_ci.py [hunk 3]
```diff
@@ -376,6 +404,11 @@ def mock_autofix(request, mocker):
 # The tests
 ##############################################################################
 
+# "GITHUB_EVENT_NAME": "",
+#     "GITHUB_EVENT_PATH": "",
+#     "GITHUB_HEAD_REF": "", #
+#     "GITHUB_REF": "",
```
Comment (brendongo) [inline cli/tests/e2e/test_ci.py]: mean to keep this?

**Review by brendongo — APPROVED** [review]

File: cli/tests/e2e/test_ci.py (hunk 3 above)
Comment (jbergler) [reply cli/tests/e2e/test_ci.py]: Nope, cleaned up.

File: cli/src/semgrep/commands/ci.py (hunk 1 above)
Comment (aryx) [reply cli/src/semgrep/commands/ci.py:113]: would have been good to add this comment in the code itself. People in 3 months will wonder why we have this os.getcwd() argument at the end.

**Comment by aryx** [comment]
thx for splitting PR in 2! I love when we first refactor to clean and then make the actual feature PR in another PR.

File: cli/tests/e2e/test_ci.py [hunk 4]
```diff
@@ -63,6 +63,34 @@
 FROZEN_ISOTIMESTAMP = "1970-01-01T00:00:00"
 
 
+# To ensure our tests are as accurate as possible, lets try to autodetect what GITHUB_ vars
+# the app code uses, so the tests can enforce the env is mocked appropriately.
```
Comment (aryx) [inline cli/tests/e2e/test_ci.py:67]: What is "the app code" in this context? Semgrep-app? the code in semgrep that interacts with the app?

File: cli/tests/e2e/test_ci.py [hunk 5]
```diff
@@ -63,6 +63,34 @@
...
+_cli_src = (Path(__file__).parent.parent.parent / "src").resolve()
+USED_GITHUB_VARS = set(
+    subprocess.run(
+        f"git grep -hPo 'GITHUB_[\\w_]*' {_cli_src}", shell=True, capture_output=True
```
Comment (aryx) [inline cli/tests/e2e/test_ci.py:71]: This currently captures the GITHUB_xxx string in cli/src/semgrep/interfaces/semgrep_output_v1.atd
which make 'make e2e' fail locally.
It's because I have
```
[submodule]
  recurse = true
  fetchJobs = 20
```
in my .gitconfig.

This git grep command seems a bit fragile to differences in environment ...
or maybe we should restrict it to the cli/src/semgrep/app/ directory?

File: cli/tests/e2e/test_ci.py (hunk 4 above)
Comment (jbergler) [reply cli/tests/e2e/test_ci.py:67]: I'm using "app" here as the generic application, ie semgrep cli

File: cli/tests/e2e/test_ci.py (hunk 5 above)
Comment (jbergler) [reply cli/tests/e2e/test_ci.py:71]: Restricting the grep might the safest, if you're ok with it - then I'd try this first though: https://github.com/returntocorp/semgrep/pull/8818
